libparsec/crates/client/src/client/recovery_device.rs

```diff
@@ -0,0 +1,373 @@
+) -> Result<(SecretKeyPassphrase, Vec<u8>), ExportRecoveryDeviceError> {
+    let (passphrase, data, recovery_device) =
+        libparsec_platform_device_loader::export_recovery_device(&client.device, device_label)
+            .await?;
+
+    // save recovery device
+    let outcome =
+        register_new_device(client.cmds.clone(), recovery_device, client.device.clone()).await?;
+
+    let latest_known_timestamps = match outcome {
+        CertificateBasedActionOutcome::LocalIdempotent => return Ok((passphrase, data)), // not sure how it could happen though
+        CertificateBasedActionOutcome::Uploaded {
+            certificate_timestamp,
+        }
+        | CertificateBasedActionOutcome::RemoteIdempotent {
+            certificate_timestamp,
+        } => PerTopicLastTimestamps::new_for_common_and_shamir(
+            certificate_timestamp,
+            certificate_timestamp,
+        ),
+    };
+    client
+        .certificates_ops
+        .poll_server_for_new_certificates(Some(&latest_known_timestamps))
+        .await
+        .map_err(|e| match e {
+            CertifPollServerError::Stopped => ExportRecoveryDeviceError::Stopped,
+            CertifPollServerError::Offline => ExportRecoveryDeviceError::Offline,
+            CertifPollServerError::InvalidCertificate(err) => {
+                ExportRecoveryDeviceError::InvalidCertificate(err)
+            }
+            CertifPollServerError::Internal(err) => err
+                .context("Cannot poll server for new certificates")
+                .into(),
+        })?;
+    Ok((passphrase, data))
+}
+
+pub async fn create_device_from_recovery(
+    cmds: Arc<AuthenticatedCmds>,
+    recovery_device: &LocalDevice,
+    device_label: &DeviceLabel,
+    save_strategy: DeviceSaveStrategy,
+    config_dir: PathBuf,
+) -> Result<AvailableDevice, ImportRecoveryDeviceError> {
+    let new_device =
+        LocalDevice::from_existing_device_for_user(recovery_device, device_label.clone());
+    // save recovery device
+    // TODO: save after upload
+    let access = {
+        let key_file = get_default_key_file(&config_dir, &new_device.device_id);
+        save_strategy.into_access(key_file)
+    };
+    let saved_device = save_device(&config_dir, &access, &new_device).await?;
+    let outcome = register_new_device(cmds, new_device, recovery_device.clone().into()).await?;
+
+    match outcome {
+        CertificateBasedActionOutcome::LocalIdempotent => return Ok(saved_device), // not sure how it could happen though
+        CertificateBasedActionOutcome::Uploaded {
+            certificate_timestamp,
+        }
+        | CertificateBasedActionOutcome::RemoteIdempotent {
+            certificate_timestamp,
+        } => PerTopicLastTimestamps::new_for_common_and_shamir(
+            certificate_timestamp,
+            certificate_timestamp,
+        ),
+    };
+
+    Ok(saved_device)
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ExportRecoveryDeviceError {
```

Comment:
Call it `ClientExportRecoveryDeviceError` since it will be used when calling `Client::export_recovery_device`